src/ratelimit_manager.cpp

```diff
@@ -0,0 +1,338 @@
+            if (rate_limit_entries.count(entry) > 0 && (rate_limit_entries.at(entry)->action == RateLimitAction::ALLOW)) {
+                rate_limit_request_counts.lookup(entry).current_requests_count_minute++;
+                rate_limit_request_counts.lookup(entry).current_requests_count_hour++;
+            }
+
+            continue;
+
+        }
+        else {
+            
+            if(!rate_limit_request_counts.contains(entry)){
+                rate_limit_request_counts.insert(entry, request_counter_t{});
+            }
+            
+            auto& request_counts = rate_limit_request_counts.lookup(entry);
+
+            // Check if last reset time was more than 1 minute ago
+            if (request_counts.last_reset_time_minute <= std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - 60) {
+                request_counts.previous_requests_count_minute = request_counts.current_requests_count_minute;
+                request_counts.current_requests_count_minute = 0;
+
+                if(request_counts.last_reset_time_minute < std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - 120) {
+                    request_counts.previous_requests_count_minute = 0;
+                }
+
+                request_counts.last_reset_time_minute = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+            }
+
+            // Check if last reset time was more than 1 hour ago
+            if (request_counts.last_reset_time_hour < std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - 3600) {
+                request_counts.previous_requests_count_hour = request_counts.current_requests_count_hour;
+                request_counts.current_requests_count_hour = 0;
+
+                if(request_counts.last_reset_time_hour < std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - 7200) {
+                    request_counts.previous_requests_count_hour = 0;
+                }
+
+                request_counts.last_reset_time_hour = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+            }
+
+            // Increment request counts
+            request_counts.current_requests_count_minute++;
+            request_counts.current_requests_count_hour++;
+
+            // Check if request count is over the limit
+            auto current_rate_for_minute = (60 - (std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - request_counts.last_reset_time_minute)) / 60  * request_counts.previous_requests_count_minute;
+            current_rate_for_minute += request_counts.current_requests_count_minute;
+
+            if(rule.throttle.minute_rate_limit >= 0 && current_rate_for_minute > rule.throttle.minute_rate_limit) {
+                if(rule.auto_ban_threshold_num >= 0 && rule.auto_ban_num_days >= 0) {
+                    temp_ban_entry_wrapped(entry, rule.auto_ban_num_days);
```

Comment:
This seems wrong. We don't have a counter for the number of times an IP / API key has been throttled. Only if __that count__ crosses `auto_ban_threshold_num` we should ban it.